pkg/cache/v3/linear.go

```diff
@@ -21,46 +21,73 @@ import (
+// TODO(valerian-roche): store serialized resource if already marshaled to compute the version.
+type cachedResource struct {
+	types.Resource
+
+	// cacheVersion is the version of the cache at the time of last update, used in sotw.
+	cacheVersion string
+	// resourceVersion is the version of the resource itself (built through stable marshaling).
+	// It is only set if computeResourceVersion is set to true on the cache.
+	resourceVersion string
+}
+
+type watches struct {
+	// sotw keeps track of current sotw watches, indexed per watch id.
+	sotw map[uint64]ResponseWatch
+	// delta keeps track of current delta watches, indexed per watch id.
+	delta map[uint64]DeltaResponseWatch
+}
+
+func newWatches() watches {
+	return watches{
+		sotw:  make(map[uint64]ResponseWatch),
+		delta: make(map[uint64]DeltaResponseWatch),
+	}
+}
+
+func (w *watches) empty() bool {
+	return len(w.sotw)+len(w.delta) == 0
+}
+
 // LinearCache supports collections of opaque resources. This cache has a
 // single collection indexed by resource names and manages resource versions
 // internally. It implements the cache interface for a single type URL and
 // should be combined with other caches via type URL muxing. It can be used to
 // supply EDS entries, for example, uniformly across a fleet of proxies.
 type LinearCache struct {
-	// Type URL specific to the cache.
+	// typeURL provides the type of resources managed by the cache.
+	// This information is used to reject requests watching another type, as well as to make
+	// decisions based on resource type (e.g. whether sotw must return full-state).
 	typeURL string
-	// Collection of resources indexed by name.
-	resources map[string]types.Resource
-	// Watches open by clients, indexed by resource name. Whenever resources
-	// are changed, the watch is triggered.
-	watches map[string]map[int64]ResponseWatch
-	// Set of watches for all resources in the collection, indexed by watch id.
-	// watch id is unique for sotw watches and is used to index them without requiring
-	// the watch itself to be hashable, as well as making logs easier to correlate.
-	watchAll map[int64]ResponseWatch
-	// Continuously incremented counter used to index sotw watches.
-	sotwWatchCount int64
-	// Set of delta watches. A delta watch always contain the list of subscribed resources
-	// together with its current version
-	// version and versionPrefix fields are ignored for delta watches, because we always generate the resource version.
-	deltaWatches map[int64]DeltaResponseWatch
-	// Continuously incremented counter used to index delta watches.
-	deltaWatchCount int64
-	// versionMap holds the current hash map of all resources in the cache when delta watches are present.
-	// versionMap is only to be used with delta xDS.
-	versionMap map[string]string
-	// Continuously incremented version.
+
+	// resources contains all resources currently set in the cache and associated versions.
+	resources map[string]cachedResource
```

Comment:
This replaces the previous `resources`, `versionMap` and `versionVector` maps that were independent (and imo confusing in naming)